security: TestRotateCerts failed

[(security).TestRotateCerts failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2634112&tab=buildLog) on [master@12c20d0f3ea71d52e7dd8277a5633cb3a8ec0e80](https://github.com/cockroachdb/cockroach/commits/12c20d0f3ea71d52e7dd8277a5633cb3a8ec0e80):

```
=== RUN   TestRotateCerts
    certs_rotation_test.go:47: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestRotateCerts184182196
E210201 19:50:15.388849 1174 (gostd) net/http/server.go:3093  [-] 1  http: TLS handshake error from 127.0.0.1:36264: remote error: tls: bad certificate
E210201 19:50:15.392552 944 1@server/server.go:2032  [n1,client=127.0.0.1:55126] 2  serving SQL client conn: remote error: tls: bad certificate
    certs_rotation_test.go:171: issuing SIGHUP
E210201 19:50:15.425584 1146 (gostd) net/http/server.go:3093  [-] 3  http: TLS handshake error from 127.0.0.1:36272: remote error: tls: bad certificate
    certs_rotation_test.go:216: structured entry for certs_reload not found in log
E210201 19:50:15.904276 1210 1@server/server.go:2032  [n1,client=127.0.0.1:55146] 4  serving SQL client conn: tls: failed to verify client certificate: x509: certificate signed by unknown authority (possibly because of "crypto/rsa: verification error" while trying to verify candidate authority certificate "Cockroach CA")
    certs_rotation_test.go:267: issuing SIGHUP
    certs_rotation_test.go:283: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestRotateCerts184182196
--- FAIL: TestRotateCerts (1.46s)

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestRotateCerts PKG=./pkg/security TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestRotateCerts.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(security).TestRotateCerts failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2639684&tab=buildLog) on [master@4736b8dd2c36b58fffbb0a665690e2c0e719d7c3](https://github.com/cockroachdb/cockroach/commits/4736b8dd2c36b58fffbb0a665690e2c0e719d7c3):

```
=== RUN   TestRotateCerts
    certs_rotation_test.go:47: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestRotateCerts891638662
E210203 07:12:05.767105 1071 (gostd) net/http/server.go:3093  [-] 1  http: TLS handshake error from 127.0.0.1:41904: remote error: tls: bad certificate
E210203 07:12:05.769353 781 1@server/server.go:2036  [n1,client=127.0.0.1:35036] 2  serving SQL client conn: remote error: tls: bad certificate
    certs_rotation_test.go:171: issuing SIGHUP
E210203 07:12:05.852211 1251 (gostd) net/http/server.go:3093  [-] 3  http: TLS handshake error from 127.0.0.1:41944: remote error: tls: bad certificate
    certs_rotation_test.go:216: structured entry for certs_reload not found in log
E210203 07:12:06.020619 1298 1@server/server.go:2036  [n1,client=127.0.0.1:35096] 4  serving SQL client conn: tls: failed to verify client certificate: x509: certificate signed by unknown authority (possibly because of "crypto/rsa: verification error" while trying to verify candidate authority certificate "Cockroach CA")
    certs_rotation_test.go:267: issuing SIGHUP
    certs_rotation_test.go:283: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestRotateCerts891638662
--- FAIL: TestRotateCerts (0.68s)

```

<details><summary>More</summary><p>
Parameters:

- TAGS=
- GOFLAGS=-parallel=4

```
make stressrace TESTS=TestRotateCerts PKG=./pkg/security TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestRotateCerts.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/security/certs_rotation_test.go
+++ b/pkg/security/certs_rotation_test.go
@@ -187,34 +187,41 @@ func TestRotateCerts(t *testing.T) {
 		})
 
 	// Check that the structured event was logged.
-	log.Flush()
-	entries, err := log.FetchEntriesFromFiles(beforeReload.UnixNano(),
-		math.MaxInt64, 10000, cmLogRe, log.WithMarkedSensitiveData)
-	if err != nil {
-		t.Fatal(err)
-	}
-	foundEntry := false
-	for _, e := range entries {
-		if !strings.Contains(e.Message, "certs_reload") {
-			continue
+	// We use SucceedsSoon here because there may be a delay between
+	// the moment SIGHUP is processed and certs are reloaded, and
+	// the moment the structured logging event is actually
+	// written to the log file.
+	testutils.SucceedsSoon(t, func() error {
+		log.Flush()
+		entries, err := log.FetchEntriesFromFiles(beforeReload.UnixNano(),
+			math.MaxInt64, 10000, cmLogRe, log.WithMarkedSensitiveData)
+		if err != nil {
+			t.Fatal(err)
 		}
-		foundEntry = true
-		// TODO(knz): Remove this when crdb-v2 becomes the new format.
-		e.Message = strings.TrimPrefix(e.Message, "Structured entry:")
-		// crdb-v2 starts json with an equal sign.
-		e.Message = strings.TrimPrefix(e.Message, "=")
-		jsonPayload := []byte(e.Message)
-		var ev eventpb.CertsReload
-		if err := json.Unmarshal(jsonPayload, &ev); err != nil {
-			t.Errorf("unmarshalling %q: %v", e.Message, err)
+		foundEntry := false
+		for _, e := range entries {
+			if !strings.Contains(e.Message, "certs_reload") {
+				continue
+			}
+			foundEntry = true
+			// TODO(knz): Remove this when crdb-v2 becomes the new format.
+			e.Message = strings.TrimPrefix(e.Message, "Structured entry:")
+			// crdb-v2 starts json with an equal sign.
+			e.Message = strings.TrimPrefix(e.Message, "=")
+			jsonPayload := []byte(e.Message)
+			var ev eventpb.CertsReload
+			if err := json.Unmarshal(jsonPayload, &ev); err != nil {
+				t.Errorf("unmarshalling %q: %v", e.Message, err)
+			}
+			if ev.Success != true || ev.ErrorMessage != "" {
+				t.Errorf("incorrect event: expected success with no error, got %+v", ev)
+			}
 		}
-		if ev.Success != true || ev.ErrorMessage != "" {
-			t.Errorf("incorrect event: expected success with no error, got %+v", ev)
+		if !foundEntry {
+			return errors.New("structured entry for certs_reload not found in log")
 		}
-	}
-	if !foundEntry {
-		t.Error("structured entry for certs_reload not found in log")
-	}
+		return nil
+	})
 
 	// Nothing changed in the first SQL client: the connection is already established.
 	if _, err := firstSQLClient.Exec("SELECT 1"); err != nil {

